### PR TITLE
FIX: add nil pointer check in dex and order

### PR DESCRIPTION
**Pull request by RevelationOfTuring**
Close：#255 

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/dex/keeper/product_lock.go [hunk 1]
```diff
@@ -25,6 +25,9 @@ func (k Keeper) getProductLock(ctx sdk.Context, product string) *ordertypes.Prod
 
 // LockTokenPair locks token pair
 func (k Keeper) LockTokenPair(ctx sdk.Context, product string, lock *ordertypes.ProductLock) {
+	if lock == nil {
+		panic("a nil pointer appears")
+	}
```
Comment (ilovers) [inline x/dex/keeper/product_lock.go:30]: Panic is dangerous, just return here?

File: x/dex/keeper/querier.go [hunk 2]
```diff
@@ -103,6 +103,9 @@ func queryDeposits(ctx sdk.Context, req abci.RequestQuery, keeper IKeeper) (res
 
 	var deposits []depositsData
 	for _, product := range tokenPairs {
+		if product == nil {
+			return res, types.ErrNilPointer()
```
Comment (zhongqiuwood) [inline x/dex/keeper/querier.go:107]: use panic

File: x/dex/keeper/querier.go [hunk 3]
```diff
@@ -133,16 +136,19 @@ func queryDeposits(ctx sdk.Context, req abci.RequestQuery, keeper IKeeper) (res
...
 	}
 
 	tokenPairs := keeper.GetTokenPairsOrdered(ctx)
 
-	var products = []string{}
+	var products []string
 
 	for _, tokenPair := range tokenPairs {
+		if tokenPair == nil {
+			return res, types.ErrNilPointer()
```
Comment (zhongqiuwood) [inline x/dex/keeper/querier.go:150]: use panic

File: x/dex/keeper/querier.go [hunk 4]
```diff
@@ -181,12 +187,15 @@ func queryProductsDelisting(ctx sdk.Context, keeper IKeeper) (res []byte, err sd
 	tokenPairs := keeper.GetTokenPairs(ctx)
 	tokenPairLen := len(tokenPairs)
 	for i := 0; i < tokenPairLen; i++ {
+		if tokenPairs[i] == nil {
+			return res, types.ErrNilPointer()
```
Comment (zhongqiuwood) [inline x/dex/keeper/querier.go:191]: use panic

File: x/dex/keeper/keeper.go [hunk 5]
```diff
@@ -202,13 +201,18 @@ func (k Keeper) updateUserTokenPair(ctx sdk.Context, product string, owner, to s
 }
 
 // UpdateTokenPair updates token pair in the store and the cache
-func (k Keeper) UpdateTokenPair(ctx sdk.Context, product string, tokenPair *types.TokenPair) {
+func (k Keeper) UpdateTokenPair(ctx sdk.Context, product string, tokenPair *types.TokenPair) sdk.Error {
+	if tokenPair == nil {
+		return types.ErrNilPointer()
```
Comment (zhongqiuwood) [inline x/dex/keeper/keeper.go:206]: use painc, and return nothing

File: x/dex/keeper/invariant.go [hunk 6]
```diff
@@ -20,6 +20,9 @@ func ModuleAccountInvariant(keeper IKeeper, supplyKeeper SupplyKeeper) sdk.Invar
 
 		// get product deposits
 		for _, product := range keeper.GetTokenPairs(ctx) {
+			if product == nil {
+				return sdk.FormatInvariant(types.ModuleName, "token pairs", "a nil pointer appears"), true
```
Comment (zhongqiuwood) [inline x/dex/keeper/invariant.go:24]: use panic, as keeper.GetTokenPairs never returns a slice containing a nil pointer

File: x/dex/keeper/keeper.go [hunk 7]
```diff
@@ -77,6 +72,10 @@ func (k Keeper) deleteUserTokenPair(ctx sdk.Context, owner sdk.AccAddress, pair
 // SaveTokenPair saves the token pair to db
 // key is base:quote
 func (k Keeper) SaveTokenPair(ctx sdk.Context, tokenPair *types.TokenPair) error {
+	if tokenPair == nil {
+		return types.ErrNilPointer()
```
Comment (zhongqiuwood) [inline x/dex/keeper/keeper.go:76]: use panic, as we never pass a nil into this method.

File: x/dex/keeper/product_lock.go [hunk 8]
```diff
@@ -25,6 +25,9 @@ func (k Keeper) getProductLock(ctx sdk.Context, product string) *ordertypes.Prod
 
 // LockTokenPair locks token pair
 func (k Keeper) LockTokenPair(ctx sdk.Context, product string, lock *ordertypes.ProductLock) {
+	if lock == nil {
+		panic("a nil pointer appears")
```
Comment (zhongqiuwood) [inline x/dex/keeper/product_lock.go:29]: change(and subsequent ones as well)：
a nil pointer appears
to
the nil pointer is not expected

**Pull request merged**